propagation.go

```diff
@@ -298,3 +304,44 @@ func (p *textMapPropagator) removeBaggageKeyPrefix(key string) string {
 	// TODO decodeBaggageHeaderKey add caching and escaping
 	return key[len(p.headerKeys.TraceBaggageHeaderPrefix):]
 }
+
+type compositePropagator struct {
+	propagators []Propagator
+}
+
+// NewCompositePropagator creates a Propagator that can be used to perform extraction and injection of SpanContexts from
+// and into carriers from multiple Propagators.
+func NewCompositePropagator(propagators []Propagator) Propagator {
+	return &compositePropagator{
+		propagators: propagators,
+	}
+}
+
+// Inject attempts to inject a SpanContext into a given carrier using the underlying propagators. If one of the
+// propagators is unable to inject the SpanContext, this method errors out (causing injection attempts using the
+// subsequent propagators to be skipped).
+func (p *compositePropagator) Inject(sc SpanContext, abstractCarrier interface{}) error {
+	for _, propagator := range p.propagators {
+		err := propagator.Inject(sc, abstractCarrier)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Extract attempts to extract a SpanContext from the given carrier using the underlying propagators. If multiple
+// propagators are capable of extracting a valid (different) SpanContext from the carrier, the first extracted
+// SpanContext will be returned.
+func (p *compositePropagator) Extract(abstractCarrier interface{}) (SpanContext, error) {
+	for _, propagator := range p.propagators {
+		sc, err := propagator.Extract(abstractCarrier)
+		if err == nil {
+			return sc, nil
+		}
+	}
+
+	// TODO: Figure out what we should return here in case of error.
```

Comment:
Choosing an error type is not so simple here as e.g. all the underlying extractor may return something like `ErrUnsupportedFormat` (in which case that would arguably be the correct error for a composite propagator).

I've chosen to return `ErrSpanContextNotFound` here to simplify (but happy to discuss).

Reply:
(See general comment).